app/controllers/spree/affirm_controller.rb

```diff
@@ -0,0 +1,37 @@
+module Spree
+  class AffirmController < Spree::StoreController
+    protect_from_forgery except: [:confirm]
+
+    def confirm
+      order = current_order || raise(ActiveRecord::RecordNotFound)
+      checkout_token = affirm_params[:checkout_token]
+
+      if !checkout_token
+        return redirect_to checkout_state_path(order.state), notice: "Invalid order confirmation data passed in"
+      end
+
+      if order.complete?
+        return redirect_to spree.order_path(order), notice: "Order is already in complete state"
+      end
+
+      affirm_checkout = Solidus::Affirm::Checkout.new(token: checkout_token)
+
+      if affirm_checkout.valid?
+        affirm_checkout.save
```

Comment:
This should all be done in a transaction. We dont want to persist the checkout if that payment creation fails, do we?